### PR TITLE
Introduce `TaskContainer.getByName<T>`

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by eskatos** [comment]
❓can't we do this on `NamedDomainObjectContainer` and the corresponding scope instead?

**Comment by bamboo** [comment]
We can't add it to `NamedDomainObjectContainer` because that would require two generic parameters. We might be able to add it to the scope but we have another issue to solve that holistically (#716).

**Comment by eskatos** [comment]
What about the following?

```kotlin
inline fun <reified T : Any> NamedDomainObjectCollection<out Any>.getByName(name: String) =
    getByName(name).let {
        it as? T
            ?: throw IllegalStateException(
                "Element '$name' of type '${it::class.java.name}' cannot be cast to '${T::class.qualifiedName}'.")
    }

inline fun <reified T : Any> NamedDomainObjectCollection<out Any>.getByName(name: String, configure: T.() -> Unit) =
        getByName<T>(name).also(configure)
```

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt [hunk 1]
```diff
@@ -25,3 +26,40 @@ import org.gradle.api.tasks.TaskContainer
...
+ *
+ * @param name task name
+ * @param configure configuration action to apply to the task before returning it
+ * @return task, never null
+ * @throws [UnknownDomainObjectException] When the given task is not found.
+ * @throws [IllegalStateException] When the given task cannot be cast to the expected type.
+ */
+@Suppress("extension_shadowed_by_member")
+inline fun <reified T : Any> TaskContainer.getByName(name: String, configure: T.() -> Unit) =
+    getByName<T>(name).also(configure)
```
Comment (JLLeitschuh) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt]: Does Kotlin still not support default parameters on inline functions?

Comment (bamboo) [reply provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt]: It does. I preferred not to mix the two concerns.

Comment (JLLeitschuh) [reply provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt]: Okay. Can you explain the mentality behind that a bit?
Cleaner code by making them two different methods?

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt [hunk 2]
```diff
@@ -25,3 +26,40 @@ import org.gradle.api.tasks.TaskContainer
...
+ * @param name task name
+ * @return task, never null
+ * @throws [UnknownDomainObjectException] When the given task is not found.
+ * @throws [IllegalStateException] When the given task cannot be cast to the expected type.
+ */
+@Suppress("extension_shadowed_by_member")
+inline fun <reified T : Any> TaskContainer.getByName(name: String) =
+    getByName(name).let {
+        it as? T
+            ?: throw IllegalStateException(
```
Comment (JLLeitschuh) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt]: Wouldn't this be an `IllegalArgumentException`? Technically `T` is an argument correct?

File: provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt [hunk 3]
```diff
@@ -25,3 +26,40 @@ import org.gradle.api.tasks.TaskContainer
...
+ * If a task with the given [name] is not found, [UnknownDomainObjectException] is thrown.
+ * If the task is found but cannot be cast to the expected type [T], [IllegalStateException] is thrown.
+ *
+ * @param name task name
+ * @return task, never null
+ * @throws [UnknownDomainObjectException] When the given task is not found.
+ * @throws [IllegalStateException] When the given task cannot be cast to the expected type.
+ */
+@Suppress("extension_shadowed_by_member")
+inline fun <reified T : Any> TaskContainer.getByName(name: String) =
```
Comment (JLLeitschuh) [inline provider/src/main/kotlin/org/gradle/kotlin/dsl/TaskContainerExtensions.kt]: Why isn't this on `TaskCollection`?

**Comment by bamboo** [comment]
We decided to handle this differently.

**Comment by bamboo** [comment]
Thanks for the review @JLLeitschuh. We're going to open a new PR soon with a different take on this.

**Comment by JLLeitschuh** [comment]
Okay. Thanks.

**Pull request closed without merging**